data/sea/51-chord.h

```diff
@@ -0,0 +1,133 @@
+#include "50-psv-error.h"
+
+#if !ICICLE_NO_PSV
+
+typedef struct {
+    const iint_t  entity_size;
+    const iint_t  count;
+    const char    entity[0];
+/*  const char    entity[entity_size + 1]; */
+/*  const idate_t dates[count];            */
+} ichord_t;
+
+typedef struct {
+    const iint_t   magic;   /* CHORDATA */
+    const iint_t   version; /* 1 */
+    const iint_t   max_chord_count;
+    const ichord_t chords[0];
+} ichord_header_t;
+
+typedef struct {
+    void           *mapped_ptr;
+    size_t          mapped_size;
+    iint_t          max_chord_count;
+    const ichord_t *chords;
+} ichord_file_t;
+
+static psv_error_t ichord_file_mmap (int fd, ichord_file_t *file)
+{
+    static ichord_file_t empty;
+    *file = empty;
+
+    if (fd == 0) {
+        file->max_chord_count = 1;
+        return 0;
+    }
+
+    struct stat stat;
+    int stat_error = fstat (fd, &stat);
+
+    if (stat_error != 0)
+        return psv_alloc_error ("failed to stat chord file", 0, 0);
+
+    size_t mapped_size = stat.st_size;
+    void  *mapped_ptr  = mmap (0, mapped_size, PROT_READ, MAP_PRIVATE, fd, 0);
+
+    if (mapped_ptr == MAP_FAILED)
+        return psv_alloc_error ("failed to memory map chord file", 0, 0);
+
+    file->mapped_ptr  = mapped_ptr;
+    file->mapped_size = mapped_size;
+
+    ichord_header_t *hdr = mapped_ptr;
+
+#if ICICLE_DEBUG
+    fprintf (stderr, "ichord_file_mmap: mapped %zu bytes to %p\n", mapped_size, mapped_ptr);
+    fprintf (stderr, "ichord_file_mmap: magic           = 0x%08llx\n", hdr->magic);
+    fprintf (stderr, "ichord_file_mmap: version         = %lld\n",     hdr->version);
+    fprintf (stderr, "ichord_file_mmap: max_chord_count = %lld\n",     hdr->max_chord_count);
+    fprintf (stderr, "ichord_file_mmap: chords          = %p\n",       hdr->chords);
+#endif
+
+    if (hdr->magic != 0x41544144524f4843 /* CHORDATA */)
+        return psv_alloc_error ("invalid magic number in chord file", 0, 0);
+
+    if (hdr->version != 1)
+        return psv_alloc_error ("only version 1 chord files are supported", 0, 0);
+
+    file->max_chord_count = hdr->max_chord_count;
+    file->chords          = hdr->chords;
+
+    return 0;
+}
+
+static psv_error_t ichord_file_unmap (const ichord_file_t *file)
+{
+    if (file->mapped_ptr == 0)
+        return 0;
+
+    int error = munmap ((void *)file->mapped_ptr, file->mapped_size);
+
+    if (error != 0)
+        return psv_alloc_error ("failed to unmap chord file", 0, 0);
+
+    return 0;
+}
+
+static const idate_t * INLINE ichord_dates (const ichord_t *chord)
+{
+    size_t      entity0_size = (size_t) chord->entity_size + 1;
+    const char* entity_end   = chord->entity + entity0_size;
+
+    return (const idate_t *)entity_end;
+}
+
+static const ichord_t * INLINE ichord_next (const ichord_t *chord)
```

Comment:
as we start writing more and more complicated C, I'd be tempted to start encoding invariants and preconditions as a Hoare logic sort of thing. doesn't have to be exact, but something like
```
{ chord /= 0 /\ VALID_CHORD_FILE }
chord' = ichord_next(chord)
{ chord' == 0 \/ chord'->entity > chord->entity }
```

just an idea for the future, I wouldn't worry about going back and adding them to this